pytorch/torchx
# Added Kueue Job Scheduler

Added a Kubernetes Kueue batch job scheduler based on the Kubernetes Scheduler
Note: variable `local_kueue="local-kueue-name"` is required in the scheduler args for the `queue-name` label and for priority add `kueue_priority_class="kueue-priority-class-name"` to the sheduler args

## Testing
Set up a Kubernetes/Openshift cluster
* Install [Kueue](https://kueue.sigs.k8s.io/docs/installation/#install-a-released-version) 
    * Create necessary Kueue resources: [Cluster Queue](https://kueue.sigs.k8s.io/docs/concepts/cluster_queue/), [Resource Flavour](https://kueue.sigs.k8s.io/docs/concepts/resource_flavor/#empty-resourceflavor), [Local Queue](https://kueue.sigs.k8s.io/docs/concepts/local_queue/) and [workload priority classes](https://kueue.sigs.k8s.io/docs/concepts/workload_priority_class/) for testing job priority
* Clone this branch locally and create a torchx whl file with `python3 setup.py bdist_wheel`
* Install the whl file with `pip install dist/torchx-0.7.0.dev0-py3-none-any.whl`
* Authenticate to your K8s cluster
* Run a sample job with ```torchx run --scheduler kueue_job --scheduler_args namespace=default,local_kueue="default-kueue",image_repo="user/alpine" utils.echo --image alpine:latest --msg hello``` - should return something like `kueue_job://torchx_user/1234`
* Get the jobs status with `torchx status kueue_job://torchx_user/1234`
* All jobs created with this scheduler should gain the `Suspended`/`JobResumed` status 

Test plan:

Created Unit tests based on Kubernetes Unit tests.
